File: sdk/python/adp_sdk/integrations/langgraph.py
"""
ADP → LangGraph and LangGraph → ADP conversion utilities.

See spec/framework-interop.md §LangGraph Mapping for the full mapping guide.
See spec/runtime-flow-binding.md for backend selection and condition expression rules.
"""
from __future__ import annotations

import json
import operator as op_module
from collections import defaultdict
from typing import Any, Callable, TypedDict

try:
    from langgraph.graph import END, StateGraph
    _AVAILABLE = True
except ImportError:
    END = None  # type: ignore[assignment]
    StateGraph = None  # type: ignore[assignment]
    _AVAILABLE = False

COMPAT_MATRIX: dict[str, list[str]] = {
    "llm":       ["openai", "anthropic", "vllm", "bedrock", "langchain", "litellm", "llm:"],
    "tool":      ["python", "docker", "http", "mcp", "wasm", "tool:"],
    "retriever": ["pinecone", "weaviate", "chroma", "redis", "pgvector", "memory:"],
    "evaluator": [],
    "router":    [],
    "input":     [],
    "output":    [],
}

_OPS: dict[str, Any] = {
    "==": op_module.eq,
    "!=": op_module.ne,
    ">":  op_module.gt,
    ">=": op_module.ge,
    "<":  op_module.lt,
    "<=": op_module.le,
}


class ADPState(TypedDict):
    inputs: dict
    context: dict[str, Any]
    memory: dict
    tool_responses: dict[str, list]


BackendFactory = Callable[[dict, dict], Callable[[ADPState], ADPState]]


def make_condition_fn(condition_str: str) -> Callable[[dict], bool]:
    """Parse an ADP condition string into a callable that evaluates against state."""
    key, op_str, value = condition_str.split(" ", 2)
    op_fn = _OPS.get(op_str)
    if op_fn is None:
        raise ValueError(f"Unsupported operator '{op_str}' in condition: {condition_str}")
    path = key.split(".")
    try:
        parsed_value = json.loads(value)
    except json.JSONDecodeError:
        parsed_value = value
    def condition_fn(state: dict) -> bool:
        current: Any = state
        for segment in path:
            current = current[segment]
        return op_fn(current, parsed_value)
    return condition_fn


def resolve_backend(node: dict, execution: list[dict]) -> dict:
    """Select the first compatible runtime.execution entry for this node kind."""
    runtime_ref = node.get("runtime_ref")
    if runtime_ref:
        entry = next((e for e in execution if e["id"] == runtime_ref), None)
        if entry is None:
            raise ValueError(f"runtime_ref '{runtime_ref}' not found in runtime.execution")
        return entry
    kind = node.get("kind", "")
    compatible = COMPAT_MATRIX.get(kind, [])
    for entry in execution:
        backend = entry.get("backend", "")
        if backend in compatible or any(backend.startswith(p) for p in compatible if p.endswith(":")):
            return entry
    return execution[0] if execution else {}


def resolve_callable(
    node: dict,
    manifest: dict,
    backend_factory: BackendFactory | None,
    entry: dict,
) -> Callable:
    """Resolve a callable for a node, handling tool_ref lookup."""
    tool_ref = node.get("tool_ref")
    if tool_ref:
        tools = manifest.get("tools", {})
        tool_entry = None
        for tool_list_key in ("mcp_servers", "http_apis", "sql_functions"):
            for tool in tools.get(tool_list_key, []):
                if tool.get("id") == tool_ref:
                    tool_entry = tool
                    break
            if tool_entry:
                break
        if tool_entry is None:
            raise ValueError(f"tool_ref '{tool_ref}' not found in tools")
        if backend_factory:
            return backend_factory(node, {"tool_entry": tool_entry, **entry})
        return _default_callable(node, entry)
    if backend_factory:
        return backend_factory(node, entry)
    return _default_callable(node, entry)


def _default_callable(node: dict, _entry: dict) -> Callable[[ADPState], ADPState]:
    """Minimal no-op callable used when no backend_factory is provided."""
    node_id = node["id"]
    kind = node.get("kind", "")
    def _fn(state: ADPState) -> ADPState:
        new_state = dict(state)
        if kind in ("llm", "retriever", "evaluator"):
            new_state["context"] = {**state.get("context", {}), node_id: {}}
        elif kind == "tool":
            responses = list(state.get("tool_responses", {}).get(node_id, []))
            new_state["tool_responses"] = {**state.get("tool_responses", {}), node_id: responses}
        return new_state  # type: ignore[return-value]
    return _fn


def build_langgraph_from_adp(
    manifest: dict,
    backend_factory: BackendFactory | None = None,
) -> tuple:
    """Build a compiled LangGraph StateGraph from an ADP manifest.

    Returns (compiled_graph, adp_node_map) where adp_node_map is a
    dict[str, dict] preserving ADP node metadata for round-trip support.
    """
    if not _AVAILABLE:
        raise ImportError("langgraph required: pip install 'adp-sdk[langgraph]'")
    adp_node_map: dict[str, dict] = {}
    graph: Any = StateGraph(ADPState)
    flow = manifest["flow"]["graph"]
    runtime = manifest["runtime"]
    execution = runtime.get("execution", [])

    for node in flow["nodes"]:
        adp_node_map[node["id"]] = node
        entry = resolve_backend(node, execution)
        fn = resolve_callable(node, manifest, backend_factory, entry)
        graph.add_node(node["id"], fn)

    cond_edges: dict[str, list[dict]] = defaultdict(list)
    for edge in flow.get("edges", []):
        if "condition" in edge:
            cond_edges[edge["from"]].append(edge)
        else:
            graph.add_edge(edge["from"], edge["to"])

    for source, edges in cond_edges.items():
        target_map = {e["to"]: make_condition_fn(e["condition"]) for e in edges}
        def _router(state: ADPState, tm: dict = target_map) -> str:
            for target_id, cond_fn in tm.items():
                if cond_fn(state):
                    return target_id
            return END
        graph.add_conditional_edges(
            source, _router, {k: k for k in target_map} | {END: END}
        )

    for start in flow["start_nodes"]:
        graph.set_entry_point(start)

    return graph.compile(), adp_node_map


def adp_from_langgraph(
    graph: Any,
    adp_node_map: dict[str, dict],
    original_manifest: dict,
) -> dict:
    """Reconstruct a partial ADP manifest from a compiled LangGraph graph.

    Structural fidelity only: node IDs, edge connections, start/end nodes.
    Non-structural ADP fields (model_ref, suite_ref, runtime_ref) are recovered
    from adp_node_map — they are not stored natively in LangGraph.
    """
    if not _AVAILABLE:
        raise ImportError("langgraph required: pip install 'adp-sdk[langgraph]'")
    draw = graph.get_graph()
    nodes = [
        adp_node_map.get(nid, {"id": nid, "kind": "unknown"})
        for nid in draw.nodes
        if nid not in ("__start__", "__end__")
    ]
    edges = [
        {"from": e.source, "to": e.target}
        for e in draw.edges
        if e.source != "__start__" and e.target != "__end__"
    ]
    start_nodes = [e.target for e in draw.edges if e.source == "__start__"]
    end_nodes = [e.source for e in draw.edges if e.target == "__end__"]
    return {
        **original_manifest,
        "flow": {
            **original_manifest.get("flow", {}),
            "graph": {
                "nodes": nodes,
                "edges": edges,
                "start_nodes": start_nodes,
                "end_nodes": end_nodes,
            },
        },
    }

File: sdk/typescript/src/integrations/langgraph.ts
/**
 * ADP → LangGraph and LangGraph → ADP conversion utilities.
 *
 * Lazy import: module is importable whether or not @langchain/langgraph is installed.
 * Throws a clear error at call time when the package is absent.
 *
 * See spec/framework-interop.md §LangGraph Mapping for the full mapping guide.
 */
import { createRequire } from "module";
import type { ADP, FlowNode } from "../adp.js";

const _require = createRequire(import.meta.url);
let _lg: any = null;
try {
  _lg = _require("@langchain/langgraph");
} catch {
  _lg = null;
}

export interface ADPState {
  inputs: Record<string, unknown>;
  context: Record<string, unknown>;
  memory: Record<string, unknown>;
  tool_responses: Record<string, unknown[]>;
}

export type BackendFactory = (node: FlowNode, entry: Record<string, unknown>) => (state: ADPState) => ADPState;

export const COMPAT_MATRIX: Record<string, string[]> = {
  llm:       ["openai", "anthropic", "vllm", "bedrock", "langchain", "litellm"],
  tool:      ["python", "docker", "http", "mcp", "wasm"],
  retriever: ["pinecone", "weaviate", "chroma", "redis", "pgvector"],
  evaluator: [],
  router:    [],
  input:     [],
  output:    [],
};

function _assertAvailable(): void {
  if (!_lg) {
    throw new Error(
      "@langchain/langgraph required: npm install @langchain/langgraph"
    );
  }
}

/* c8 ignore start */
function parseCondition(conditionStr: string): (state: ADPState) => boolean {
  const parts = conditionStr.split(" ");
  if (parts.length < 3) throw new Error(`Invalid condition: ${conditionStr}`);
  const [keyPath, op, rawValue] = [parts[0], parts[1], parts.slice(2).join(" ")];
  const path = keyPath.split(".");
  let value: unknown;
  try { value = JSON.parse(rawValue); } catch { value = rawValue; }
  const ops: Record<string, (a: unknown, b: unknown) => boolean> = {
    "==": (a, b) => a === b,
    "!=": (a, b) => a !== b,
    ">":  (a, b) => (a as number) > (b as number),
    ">=": (a, b) => (a as number) >= (b as number),
    "<":  (a, b) => (a as number) < (b as number),
    "<=": (a, b) => (a as number) <= (b as number),
  };
  const opFn = ops[op];
  if (!opFn) throw new Error(`Unsupported operator '${op}' in condition: ${conditionStr}`);
  return (state: ADPState) => {
    let current: unknown = state;
    for (const segment of path) {
      current = (current as Record<string, unknown>)[segment];
    }
    return opFn(current, value);
  };
}

function _defaultCallable(node: FlowNode): (state: ADPState) => ADPState {
  const nodeId = node.id;
  const kind = node.kind;
  return (state: ADPState): ADPState => {
    const newState = { ...state };
    if (kind === "llm" || kind === "retriever" || kind === "evaluator") {
      newState.context = { ...state.context, [nodeId]: {} };
    } else if (kind === "tool") {
      const responses = [...(state.tool_responses[nodeId] ?? [])];
      newState.tool_responses = { ...state.tool_responses, [nodeId]: responses };
    }
    return newState;
  };
}
/* c8 ignore stop */

export function makeConditionFn(conditionStr: string): (state: ADPState) => boolean {
  return parseCondition(conditionStr);
}

/* c8 ignore start */
export function buildLangGraphFromAdp(
  manifest: ADP,
  backendFactory?: BackendFactory,
): { graph: unknown; nodeMap: Record<string, FlowNode> } {
  _assertAvailable();
  const { StateGraph, END } = _lg!;
  const flow = (manifest as any).flow.graph;
  const runtime = (manifest as any).runtime ?? {};
  const execution: unknown[] = runtime.execution ?? [];
  const nodeMap: Record<string, FlowNode> = {};

  const graph = new StateGraph({ channels: {} as any });

  for (const node of flow.nodes as FlowNode[]) {
    nodeMap[node.id] = node;
    const fn = backendFactory ? backendFactory(node, {}) : _defaultCallable(node);
    graph.addNode(node.id, fn as any);
  }

  const condEdges: Record<string, Array<{ from: string; to: string; condition: string }>> = {};
  for (const edge of flow.edges as Array<{ from: string; to: string; condition?: string }>) {
    if (edge.condition) {
      (condEdges[edge.from] ??= []).push(edge as any);
    } else {
      graph.addEdge(edge.from as any, edge.to as any);
    }
  }

  for (const [source, edges] of Object.entries(condEdges)) {
    const targetMap = Object.fromEntries(
      edges.map((e) => [e.to, makeConditionFn(e.condition)])
    );
    graph.addConditionalEdges(
      source as any,
      (state: ADPState) => {
        for (const [targetId, condFn] of Object.entries(targetMap)) {
          if (condFn(state)) return targetId;
        }
        return END;
      },
      { ...Object.fromEntries(Object.keys(targetMap).map((k) => [k, k])), [END]: END },
    );
  }

  for (const startId of flow.start_nodes as string[]) {
    graph.setEntryPoint(startId);
  }

  return { graph: graph.compile(), nodeMap };
}

export function adpFromLangGraph(
  graph: unknown,
  nodeMap: Record<string, FlowNode>,
  originalManifest: ADP,
): ADP {
  _assertAvailable();
  const draw = (graph as any).getGraph();
  const nodes = (draw.nodes as string[])
    .filter((nid) => nid !== "__start__" && nid !== "__end__")
    .map((nid) => nodeMap[nid] ?? { id: nid, kind: "unknown" as const });
  const edges = (draw.edges as Array<{ source: string; target: string }>)
    .filter((e) => e.source !== "__start__" && e.target !== "__end__")
    .map((e) => ({ from: e.source, to: e.target }));
  const startNodes = (draw.edges as Array<{ source: string; target: string }>)
    .filter((e) => e.source === "__start__").map((e) => e.target);
  const endNodes = (draw.edges as Array<{ source: string; target: string }>)
    .filter((e) => e.target === "__end__").map((e) => e.source);
  return {
    ...originalManifest,
    flow: {
      ...(originalManifest as any).flow,
      graph: { nodes, edges, start_nodes: startNodes, end_nodes: endNodes },
    },
  } as ADP;
}
/* c8 ignore stop */
